src/compiler/checker.ts

```diff
@@ -9544,7 +9604,7 @@ namespace ts {
                     return getUnionType([leftType, rightType]);
                 case SyntaxKind.EqualsToken:
                     checkAssignmentOperator(rightType);
-                    return rightType;
+                    return getRegularTypeOfObjectLiteral(rightType);

```

Comment:
Hmm. Contextual typing only comes from the first assignment target, so we're currently consistent with that. I suppose you could argue we should check against a union of *all* assignment targets, i.e. as long as each property is known in *some* assignment target we're good. But that would add a bunch of complexity that I don't think is justified.